Unnest array of rows based on unnestArrayOfRows

aditi-pandit opened the pull request:
Velox unnests array of row type into a single column of the
element row type. Presto also supports unnesting such arrays
into multiple columns one for each child type of element row
type. Adds unnestArrayOfRows to add such unnesting support
in Unnest PlanNode.

Needed for https://github.com/prestodb/presto/issues/20643

mbasmanova commented on velox/exec/Unnest.cpp:
```diff
@@ -220,13 +220,52 @@ RowVectorPtr Unnest::generateOutput(
+        auto rowIsIdentity = rowDecoded.isIdentityMapping();
+        auto numRowElements = rowDecoded.size();
+
+        // These need to be constructed because the DecodedVector indices/nulls
```
Would you check if DecodedVector::wrap API can be used here?

aditi-pandit replied:
Thanks @mbasmanova. DecodedVector.wrap works as well.

aditi-pandit commented:
@mbasmanova : Updated the code. PTAL.

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -1897,29 +1897,60 @@ class LimitNode : public PlanNode {
       std::vector<FieldAccessTypedExprPtr> unnestVariables,
       const std::vector<std::string>& unnestNames,
       const std::optional<std::string>& ordinalityName,
+      bool legacyUnnest,
```
> legacyUnnest

Unlike Presto, Velox is a relatively new, hence, it is a bit strange to see a parameter named 'legacyUnnest' here. Also, this name is not very descriptive. 

I wonder if we can come up with a better name.

aditi-pandit replied:
Using unnestArrayOfRows now to control the output instead.

mbasmanova commented on velox/core/PlanNode.cpp:
```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
-    } else if (variable->type()->isMap()) {
-      const auto& mapType = variable->type()->asMap();
-
+    auto type = variable->type();
```
const &

mbasmanova commented on velox/core/PlanNode.cpp:
```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
-
+    auto type = variable->type();
+    if (type->isArray()) {
+      auto elementType = type->asArray().elementType();
```
const &

mbasmanova commented on velox/core/PlanNode.cpp:
```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
+        // column for each child type of the row.
+        // Note : In legacy unnest however, array of rows like other arrays
+        // are flattened to a single column of the array type.
+        auto rowType = asRowType(elementType);
```
not need for asRowType; Type base class have size() and childAt() APIs, hence, you can use elementType directly.

mbasmanova commented on velox/core/PlanNode.cpp:
```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
+        // are flattened to a single column of the array type.
+        auto rowType = asRowType(elementType);
+        auto numChildren = rowType->size();
+        for (auto j = 0; j < numChildren; j++) {
```
j -> i

mbasmanova commented on velox/core/PlanNode.cpp:
```diff
@@ -845,25 +847,54 @@ UnnestNode::UnnestNode(
       VELOX_FAIL(
           "Unexpected type of unnest variable. Expected ARRAY or MAP, but got {}.",
-          variable->type()->toString());
+          type->toString());
```
Can you also change this to VELOX_USER_FAIL?

mbasmanova commented on velox/core/PlanNode.cpp:
```diff
@@ -895,6 +926,8 @@ folly::dynamic UnnestNode::serialize() const {
     obj["ordinalityName"] = outputType()->names().back();
   }
+
+  obj["legacyUnnest"] = legacyUnnest_;
```
We need to include this property in the output of UnnestNode::toString.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -665,9 +667,11 @@ number starting with 1.
    * - ordinalityName
      - Optional name for the ordinality column.
+   * - legacyUnnest
+     - Array of rows are typically expanded to multiple columns, one for each child type of the row. However, with legacyUnnest, array of rows (like other array types) are expanded to a single column of the row type.
```
perhaps, legacyUnnest -> flattenArrayOfRow or unnestArrayOfRow

aditi-pandit replied:
@mbasmanova : legacyUnnest unnests array of rows as a single column of row type. (!legacyUnnest) unnests as multiple columns of the child row types. So I presume you are suggesting flattenArrayOfRow or unnestArrayOfRow for the latter behavior (!legacyUnnest) (Your comment says replace legacyUnnest with those)

mbasmanova commented on velox/exec/Unnest.cpp:
```diff
@@ -182,8 +183,7 @@ RowVectorPtr Unnest::generateOutput(
 
-    auto nulls =
-        AlignedBuffer::allocate<bool>(numElements, pool(), bits::kNotNull);
+    auto nulls = allocateNulls(numElements, pool());
```
Nice refactor.

mbasmanova commented on velox/exec/Unnest.cpp:
```diff
@@ -220,13 +220,35 @@ RowVectorPtr Unnest::generateOutput(
-                elementIndices,
-                unnestBaseArray->elements(),
-                nulls);
+      auto elements = unnestBaseArray->elements();
```
I noticed that this method is very long. Perhaps, refactor to extract some helper methods.

aditi-pandit replied:
Created https://github.com/facebookincubator/velox/pull/8077 for the overall refactoring. I will move this logic into it.

mbasmanova commented on velox/exec/tests/utils/PlanBuilder.h:
```diff
@@ -890,10 +892,16 @@ class PlanBuilder {
       const std::vector<std::string>& unnestColumns,
-      const std::optional<std::string>& ordinalColumn = std::nullopt);
+      const std::optional<std::string>& ordinalColumn = std::nullopt,
+      const std::optional<bool>& legacyUnnest = std::nullopt);
```
Why do we need std::optional here? Can we just use bool?

mbasmanova commented on velox/exec/tests/utils/PlanBuilder.cpp:
```diff
@@ -1430,11 +1432,22 @@ PlanBuilder& PlanBuilder::unnest(
+        auto row = asRowType(input->childAt(0));
+        auto size = row->size();
+        for (auto i = 0; i < size; i++) {
+          unnestNames.push_back(row->nameOf(i) + "_e");
```
Should we include 'name' as well? What if there are multiple columns of the same array(row) type?

mbasmanova commented on velox/exec/tests/UnnestTest.cpp:
```diff
@@ -39,6 +39,39 @@ TEST_F(UnnestTest, basicArray) {
   auto op = PlanBuilder().values({vector}).unnest({"c0"}, {"c1"}).planNode();
   assertQuery(op, "SELECT c0, UNNEST(c1) FROM tmp WHERE c0 % 7 > 0");
+
+  // Test with array containing dictionary encoded elements.
```
This test is pretty hard to read and maintain. If we want to test encodings, it is better add a test that uses Fuzzer to generate inputs.

mbasmanova commented on velox/exec/tests/UnnestTest.cpp:
```diff
@@ -103,6 +136,160 @@ TEST_F(UnnestTest, arrayWithOrdinality) {
+  assertQuery(op, expected);
+}
+
+TEST_F(UnnestTest, encodedArrayOfRows) {
```
Let's write a fuzzer-based test. Use Fuzzer to generate inputs, then run Unnest plan on these and on flattened copies. Compare the results.

mbasmanova reviewed (commented):
@aditi-pandit Aditi, thank you for working on this. Overall looks good % a comment about parameter name, another comment about output of UnnestNode::toString() and some comments about tests.

aditi-pandit commented:
@mbasmanova : Have addressed all the comments. PTAL. https://github.com/facebookincubator/velox/pull/8077 is separated for refactoring the long method in Unnest logic pointed in review.

mbasmanova commented:
@aditi-pandit CI is red. Would you try rebasing?

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -1897,29 +1897,60 @@ class LimitNode : public PlanNode {
+  /// The output names must appear in the same order as unnestVariables.
   /// @param ordinalityName Optional name for the ordinality columns. If not
   /// present, ordinality column is not produced.
+  /// @param unnestArrayOfRows Expands array of rows to multiple columns, one
```
nit: If true, expands .... Otherwise, ...

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -1897,29 +1897,60 @@ class LimitNode : public PlanNode {
+            unnestVariables,
+            unnestNames,
+            ordinalityName,
+            true,
```
Curious if default should be false. Doesn't current logic match unnestArrayOfRows = false?

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -665,9 +667,11 @@ number starting with 1.
      - Input columns of type array or map to expand.
    * - unnestNames
-     - Names to use for expanded columns. One name per array column. Two names per map column.
+     - Names to use for expanded columns. One name per array column. Two names per map column. If array of rows (and not legacyUnnest), then one column per child type of the row.
```
> legacyUnnest

Should we update this?

mbasmanova commented on velox/exec/tests/PlanNodeToStringTest.cpp:
```diff
@@ -560,6 +560,17 @@ TEST_F(PlanNodeToStringTest, unnest) {
+
+  ASSERT_EQ("-- Unnest\n", plan->toString());
+  ASSERT_EQ(
+      "-- Unnest[a0, unnest arrays of rows] -> c1:INTEGER, a0_e:SMALLINT\n",
```
Should this example use array(row) and not array(smallint)?

mbasmanova commented on velox/exec/tests/QueryAssertionsTest.cpp:
```diff
@@ -493,4 +494,23 @@ TEST_F(QueryAssertionsTest, intervalDayTime) {
   assertQuery(plan, "SELECT * FROM tmp");
 }
 
+TEST_F(QueryAssertionsTest, equalPlans) {
```
What is this test for?

mbasmanova commented on velox/exec/tests/UnnestTest.cpp:
```diff
@@ -39,6 +40,19 @@ TEST_F(UnnestTest, basicArray) {
+
+  // Tests with an array with dictionary encoded elements.
+  VectorFuzzer::Options opts;
+  opts.containerVariableLength = false;
```
Why false? Usually, these kinds of tests have a loop with 100 or so iterations and allow Fuzzer to generate whatever encoding it chooses using fuzzInputRow.

mbasmanova commented on velox/exec/tests/UnnestTest.cpp:
```diff
@@ -103,6 +117,146 @@ TEST_F(UnnestTest, arrayWithOrdinality) {
+
+  // Test with array of rows (one of which is a constant). Each array is
+  // encoded.
+  vector_size_t size = array1.size();
```
Let's use fuzzer.

mbasmanova commented on velox/exec/tests/UnnestTest.cpp:
```diff
@@ -103,6 +117,146 @@ TEST_F(UnnestTest, arrayWithOrdinality) {
+        .planNode();
+  };
+
+  auto noUnnestPlan = [&](const VectorPtr& child1,
```
We should use the same plans but first use data generated by Fuzzer, then use same data but flattened. The results should be the same.

mbasmanova commented on velox/exec/tests/utils/PlanBuilder.cpp:
```diff
@@ -1426,8 +1426,10 @@ PlanBuilder& PlanBuilder::nestedLoopJoin(
     const std::vector<std::string>& unnestColumns,
-    const std::optional<std::string>& ordinalColumn) {
+    const std::optional<std::string>& ordinalColumn,
+    const bool unnestArrayOfRows) {
```
remove const

mbasmanova commented on velox/exec/tests/utils/QueryAssertions.h:
```diff
@@ -201,6 +201,11 @@ bool assertEqualResults(
     const std::vector<RowVectorPtr>& actual);
 
+/// Ensure both plans have the same results.
+bool assertEqualResults(
```
Maybe extract this change into a separate PR.

aditi-pandit replied:
Sure. https://github.com/facebookincubator/velox/pull/8134
